allennlp/modules/elmo.py

```diff
@@ -0,0 +1,422 @@
+import json
+
+from typing import Union, List, Dict, Any
+
+import torch
+from torch.autograd import Variable
+
+import numpy
+import h5py
+
+from overrides import overrides
+
+from allennlp.common.file_utils import cached_path
+from allennlp.common.checks import ConfigurationError
+from allennlp.common import Registrable, Params
+from allennlp.modules.elmo_lstm import ElmoLstm
+from allennlp.modules.highway import Highway
+from allennlp.modules.scalar_mix import ScalarMix
+from allennlp.nn.util import remove_sentence_boundaries, add_sentence_boundary_token_ids
+from allennlp.data import Vocabulary
+from allennlp.data.token_indexers.elmo_indexer import ELMoCharacterMapper
+
+# pylint: disable=attribute-defined-outside-init
+
+
+@Registrable.register('elmo')
+class Elmo(torch.nn.Module, Registrable):
+    """
+    Compute ELMo representations using a pre-trained bidirectional language model.
+
+    See "Deep contextualized word representations", Peters et al. for details.
+
+    This module takes character id input and computes ``num_elmo_layers`` different layers
+    of ELMo representations.  Typically ``num_elmo_layers`` is 1 or 2.  For example, in
+    the case of the SRL model in the above paper, ``num_elmo_layers=1`` where ELMo was included at
+    the input token representation layer.  In the case of the SQuAD model, ``num_elmo_layers=2``
+    as ELMo was also included at the GRU output layer.
+
+    In the implementation below, we learn separate scalar weights for each output layer,
+    but only run the biLM once on each input sequence for efficiency.
+
+    Parameters
+    ----------
+    options_file : str
+        ELMo JSON options file
+    weight_file : str
+        ELMo hdf5 weight file
+    num_elmo_layers: int
+        The number of ELMo representation layers to output.
+    do_layer_norm: bool
+        Should we apply layer normalization (passed to ``ScalarMix``)?
+    """
+    def __init__(self,
+                 options_file: str,
+                 weight_file: str,
+                 num_elmo_layers: int,
+                 do_layer_norm: bool = False) -> None:
+        super(Elmo, self).__init__()
+
+        self._elmo_lstm = _ElmoBiLm(options_file, weight_file)
+
+        self._scalar_mixes: Any = []
+        for k in range(num_elmo_layers):
+            scalar_mix = ScalarMix(self._elmo_lstm.num_layers, do_layer_norm=do_layer_norm)
+            self.add_module('scalar_mix_{}'.format(k), scalar_mix)
+            self._scalar_mixes.append(scalar_mix)
+
+    def forward(self,    # pylint: disable=arguments-differ
+                inputs: torch.Tensor) -> Dict[str, Union[torch.Tensor, List[torch.Tensor]]]:
+        """
+        Parameters
+        ----------
+        inputs: ``torch.autograd.Variable``
+            Shape ``(batch_size, timesteps, 50)`` of character ids representing the current batch.
+
+        Returns
+        -------
+        Dict with keys:
+
+        ``'elmo'``: ``List[torch.autograd.Variable]``
```

Comment:
Nit: can these be called `elmo_representations` or something?

Reply:
done